### ISSUE TITLE
Issues with $select string templates and arrays

### ISSUE THREAD

**Issue by drcmda** [revue/revue#20]
In Redux it is common to pass IDs to components so that they can observe themselves. In Revue we tried that but it looks like it can't understand array syntax (path[id]). But we can't chain them (path.id) because id's are guids. path.763726-43ad34-4334ea would be invalid Javascript.

Is there any reason why Revue uses string templates instead of real, functional selectors?

Revue:

    export default {
        props: ['id'],
        data () {
            return {
                // This won't work for keys that can't be represented as variables
                item: this.$select(`elements.${this.id} as item`)
            }
        }

React-Redux:

    @connect((state, props) => ({ item: state.elements[props.id] }))
    class Element extends React.Component {
        ...

**Comment by egoist** [comment]
because we need to know which prop you're binding to in order to update value in the vm.

**Comment by drcmda** [comment]
I was asked if if i found a solution, and i'll just drop my solution here, works more or less the same as in react redux or all the other redux bindings:

store.js

``` javascript
import cloneDeep from 'lodash/cloneDeep';
import { createStore, combineReducers } from 'redux';
export store = createStore(combineReducers({ ...reducers }));

export const connect = (selector, props) => {
    let state = undefined;
    return {
        data: () => ({ state: { } }),
        created() {
            state = selector(store.getState(), this);
            this.state = Object.keys(state).reduce((prev, key, index) =>
                ({ ...prev, [key]: cloneDeep(state[key]) }), {})
            this.unsubscribe = store.subscribe(() => {
                let next = selector(store.getState(), this);
                Object.keys(state).forEach(key => {
                    if (state[key] !== next[key]) {
                        this.$set(`state.${key}`, cloneDeep(next[key]));
                    }
                });
                state = next;
            });
        },
        beforeDestroy() {
            this.unsubscribe();
        }
    }
}
```

Component.vue

``` javascript
<template>
    <button @click="handleClick">{{ state.item.name }}</button>
</template>

import { store, actions, connect } from '../store.js';

export default {
    mixins: [connect((state, scope) => ({ item: state.elements[scope.id] }))],
    props: ['id'],
    methods: {
        handleClick(event) {
            store.dispatch(actions.event(this.id, event))
        }
    }
}
```

**Comment by callmetwan** [comment]
@drcmda I did some more testing with Revue and I didn't have any problems with using GUIDs. I realize you have a working solution so you likely have no motivation to re-test this but figured I'd throw it out there for anyone else reading this.

For what it is worth, Internally it seems Revue handles accessing all properties using array syntax (https://github.com/revue/revue/blob/master/src/index.js#L20).

**Comment by drcmda** [comment]
@callmetwan 

    let a = { 123: "abc" }
    a.123

should throw "Uncaught SyntaxError: Unexpected number." `a[123]` and `a["123"]` are ok.

**Comment by callmetwan** [comment]
I understand JavaScript's limitations on using dot notation, I was saying
I'm not able to reproduce Revue throwing an error using an integer in a
select statement.

this.$select('weights.1234 as weights')

That has no issues for me

**Comment by drcmda** [comment]
Yes, as i said, i didn't even think of trying that. I went straight to weights[1234]. Though i still don't get it, what is the "as weights" ? Anyway, the mix-in above works the same as all the other redux bindings for react, angular, ember, etc. Vues is the only one that had the odd string syntax.

**Comment by callmetwan** [comment]
To be clear, this string syntax is specific to Revue, not Vue. The `as weights` tells Revue what the name of the Vue data property we are binding the $select() to; it's like binding to scope in your example. By using a string rather than a direct property reference this allows Revue to bind to properties that may not exist in Redux yet.
